tests/test_utils.py

```diff
@@ -486,3 +492,167 @@ def template(x):
+
+        inv_ops = pu.inv(dummy_template([0, 1, 2]))
+
+        for inv_op, exp_op in zip(inv_ops, inv_queue):
+            assert inv_op.name == exp_op.name
+            assert inv_op.wires == exp_op.wires
+            assert inv_op.params == exp_op.params
+
+    def test_double_inversion(self):
+        """Test that inverting twice changes nothing."""
+        op_queue = [qml.PauliX(0), qml.PauliY(0), qml.PauliZ(0)]
+
+        inv_inv_ops = pu.inv(pu.inv(op_queue))
+
+        for inv_inv_op, exp_op in zip(inv_inv_ops, op_queue):
+            assert inv_inv_op.name == exp_op.name
+            assert inv_inv_op.wires == exp_op.wires
+            assert inv_inv_op.params == exp_op.params
+
+    def test_template_double_inversion(self):
+        """Test that inverting twice changes nothing for a template."""
+        inv_inv_ops = pu.inv(pu.inv(dummy_template([0, 1, 2])))
+
+        for inv_inv_op, exp_op in zip(inv_inv_ops, dummy_template([0, 1, 2])):
+            assert inv_inv_op.name == exp_op.name
+            assert inv_inv_op.wires == exp_op.wires
+            assert inv_inv_op.params == exp_op.params
+
+    def test_inversion_with_context(self):
+        """Test that a sequence of operations is properly inverted when a context is present."""
+        with pu.OperationRecorder() as rec:
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            pu.inv([qml.RX(1, wires=[0]), qml.RY(2, wires=[0]), qml.RZ(3, wires=[0])])
+            qml.CNOT(wires=[0, 1])
+            qml.Hadamard(wires=[0])
+
+        inv_queue = [
+            qml.Hadamard(wires=[0]),
+            qml.CNOT(wires=[0, 1]),
+            qml.RZ(3, wires=[0]).inv(),
+            qml.RY(2, wires=[0]).inv(),
+            qml.RX(1, wires=[0]).inv(),
+            qml.CNOT(wires=[0, 1]),
+            qml.Hadamard(wires=[0]),
+        ]
+
+        for inv_op, exp_op in zip(rec.queue, inv_queue):
+            assert inv_op.name == exp_op.name
+            assert inv_op.wires == exp_op.wires
+            assert inv_op.params == exp_op.params
+
+    def test_template_inversion_with_context(self):
+        """Test that a template is properly inverted when a context is present."""
+        with pu.OperationRecorder() as rec:
+            qml.Hadamard(wires=[0])
+            qml.CNOT(wires=[0, 1])
+            pu.inv(dummy_template([0, 1, 2]))
+            qml.CNOT(wires=[0, 1])
+            qml.Hadamard(wires=[0])
+
+        inv_queue = [
+            qml.Hadamard(wires=[0]),
+            qml.CNOT(wires=[0, 1]),
+            *inverted_dummy_template_operations([0, 1, 2]),
+            qml.CNOT(wires=[0, 1]),
+            qml.Hadamard(wires=[0]),
+        ]
+
+        for inv_op, exp_op in zip(rec.queue, inv_queue):
+            assert inv_op.name == exp_op.name
+            assert inv_op.wires == exp_op.wires
+            assert inv_op.params == exp_op.params
+
+    def test_inversion_with_qnode(self):
```

Comment:
The following example seems to result in an error for me:
```python
template_list = [qml.RX(0.4, wires=[1]), qml.RX(0.4, wires=[2]), qml.RX(0.4, wires=[3]), qml.RY(0.4, wires=[1]), qml.RY(0.4, wires=[2]), qml.RY(0.4, wires=[3])]

@qml.qnode(dev)
def circuit():
    pu.inv(template_list)
    return qml.expval(qml.PauliZ(wires=[0]))
```
The traceback is (with the first line being the content of the context):
```python
[<pennylane.ops.qubit.RX object at 0x7f3e7f532710>, <pennylane.ops.qubit.RX object at 0x7f3e7f532750>, <pennylane.ops.qubit.RX object at 0x7f3e7f42c090>, <pennylane.ops.qubit.RY object at 0x7f3e7f3b5a50>, <pennylane.ops.qubit.RY object at 0x7f3e7f3b5b50>, <pennylane.ops.qubit.RY object at 0x7f3e7f3b59d0>]
Traceback (most recent call last):
  File "example_template_inv.py", line 29, in <module>
    print(circuit())
  File "/pennylane/pennylane/interfaces/autograd.py", line 47, in __call__
    return self.evaluate(args, kwargs)
  File "/anaconda3/envs/py_env_1/lib/python3.7/site-packages/autograd/tracer.py", line 48, in f_wrapped
    return f_raw(*args, **kwargs)
  File "/pennylane/pennylane/qnodes/base.py", line 570, in evaluate
    self._construct(args, kwargs)
  File "/pennylane/pennylane/qnodes/jacobian.py", line 60, in _construct
    super()._construct(args, kwargs)
  File "/pennylane/pennylane/qnodes/base.py", line 355, in _construct
    res = self.func(*arg_vars, **kwargs)
  File "example_template_inv.py", line 26, in circuit
    pu.inv(list1)
  File "/pennylane/pennylane/utils.py", line 378, in inv
    indices = [qml._current_context.queue.index(op) for op in operation_list]
  File "/pennylane/pennylane/utils.py", line 378, in <listcomp>
    indices = [qml._current_context.queue.index(op) for op in operation_list]
ValueError: <pennylane.ops.qubit.RX object at 0x7f3e7a76b410> is not in list
```

Reply:
Thanks for getting that one! I actually assumed that all ops that are given are already in the queue if inside a QNode. I fixed that, now it will work.